File: leetcode/0061.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        if (head == nullptr || head->next == nullptr)
        {
            return head;
        }
        int n = 1;
        ListNode *cur = head;
        while (cur->next != nullptr)
        {
            n++;
            cur = cur->next;
        }
        int pre = n - k % n;
        if (pre == n)
        {
            return head;
        }
        cur->next = head;
        while (pre--)
        {
            cur = cur->next;
        }
        ListNode *ans = cur->next;
        cur->next = nullptr;
        return ans;
    }
};

File: leetcode/0061.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	n, cur := 1, head
	for ; cur.Next != nil; n++ {
		cur = cur.Next
	}
	pre := n - k%n
	if pre == n {
		return head
	}
	cur.Next = head
	for ; pre > 0; pre-- {
		cur = cur.Next
	}
	ans := cur.Next
	cur.Next = nil
	return ans
}
